From 15c2630d703d3522b188a276d21f464587af66fd Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Mon, 25 Aug 2014 07:52:26 -0400
Subject: [tools] lib/traceevent: Check return value of realloc()

Message-id: <1408953210-26343-259-git-send-email-jolsa@redhat.com>
Patchwork-id: 89041
O-Subject: [PATCH RHEL7.1 BZ1133083 258/322] tools lib traceevent: Check return value of realloc()
Bugzilla: 1133083
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1133083
https://bugzilla.redhat.com/show_bug.cgi?id=1133083

upstream
========
commit 3026bba3c37711234771349ca020d9a85e572f60
Author: Namhyung Kim <namhyung@kernel.org>
Date: Wed Jan 15 10:45:25 2014 +0900

description
===========
If realloc() fails, it'll leak the buffer.  Also increate buffer size
only if the allocation succeeded.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/lib/traceevent/trace-seq.c b/tools/lib/traceevent/trace-seq.c
index f711213..e454a2c 100644
--- a/tools/lib/traceevent/trace-seq.c
+++ b/tools/lib/traceevent/trace-seq.c
@@ -91,11 +91,16 @@ void trace_seq_destroy(struct trace_seq *s)
 
 static void expand_buffer(struct trace_seq *s)
 {
- s->buffer_size += TRACE_SEQ_BUF_SIZE;
- s->buffer = realloc(s->buffer, s->buffer_size);
- if (WARN_ONCE(!s->buffer,
-        "Can't allocate trace_seq buffer memory"))
+ char *buf;
+
+ buf = realloc(s->buffer, s->buffer_size + TRACE_SEQ_BUF_SIZE);
+ if (WARN_ONCE(!buf, "Can't allocate trace_seq buffer memory")) {
   s->state = TRACE_SEQ__MEM_ALLOC_FAILED;
+  return;
+ }
+
+ s->buffer = buf;
+ s->buffer_size += TRACE_SEQ_BUF_SIZE;
 }
 
 /**
-- 
1.7.1